src/coreComponents/constitutive/fluid/MultiPhaseMultiComponentFluid.cpp

```diff
@@ -19,418 +19,214 @@
+                  InputError );
+
+  // NOTE: for now, the names of the phases are still hardcoded here
+  // Later, we could read them from the XML file and we would then have a general class here
+
+  char const * expectedWaterPhaseNames[] = { "Water", "water", "Liquid", "liquid" };
+  m_p1Index = PVTFunctionHelpers::findName( m_phaseNames,
+                                            expectedWaterPhaseNames );
+  GEOSX_THROW_IF( m_p1Index < 0 || m_p1Index >= m_phaseNames.size(),
+                  "Phase water/liquid is not found!",
+                  InputError );
+
+  char const * expectedGasPhaseNames[] = { "CO2", "co2", "gas", "Gas" };
+  m_p2Index = PVTFunctionHelpers::findName( m_phaseNames,
+                                            expectedGasPhaseNames );
+  GEOSX_THROW_IF( m_p2Index < 0 || m_p2Index >= m_phaseNames.size(),
+                  "Phase co2/gas is not found!",
+                  InputError );
 
   createPVTModels();
-
-}
-
-void MultiPhaseMultiComponentFluid::initializePostSubGroups()
-{
-  MultiFluidBase::initializePostSubGroups();
-
-  //  CreatePVTModels();
-
 }
 
-
-void MultiPhaseMultiComponentFluid::createPVTModels()
+template< typename P1DENS, typename P1VISC, typename P2DENS, typename P2VISC, typename FLASH >
+void MultiPhaseMultiComponentFluid< P1DENS, P1VISC, P2DENS, P2VISC, FLASH >::createPVTModels()
 {
-  for( string & filename : m_phasePVTParaFiles )
+  // 1) Create the viscosity and density models
+  for( string const & filename : m_phasePVTParaFiles )
   {
     std::ifstream is( filename );
-
     constexpr std::streamsize buf_size = 256;
     char buf[buf_size];
 
-    while( is.getline( buf, buf_size ))
+    // NOTE: for now, the names of the models are still hardcoded here
+    // Later, we could read them from the XML file and we would then have a general class here
+
+    while( is.getline( buf, buf_size ) )
     {
       string const str( buf );
       string_array const strs = stringutilities::tokenize( str, " " );
 
       if( strs[0] == "DensityFun" )
       {
-        m_phaseDensityFuns.emplace_back( PVTFunction::CatalogInterface::factory( strs[ 1 ],
-                                                                                 strs,
-                                                                                 m_componentNames,
-                                                                                 m_componentMolarWeight ) );
+        if( strs[1] == "BrineCO2Density" )
+        {
+          m_p1Density = std::make_unique< P1DENS >( strs, m_componentNames, m_componentMolarWeight );
+          m_p1DensityWrapper.emplace_back( m_p1Density->createKernelWrapper() );
+        }
+        else if( strs[1] == "SpanWagnerCO2Density" )
+        {
+          m_p2Density = std::make_unique< P2DENS >( strs, m_componentNames, m_componentMolarWeight );
+          m_p2DensityWrapper.emplace_back( m_p2Density->createKernelWrapper() );
+        }
```

Comment:
Can we use something like `P1DENS::catalogName()` and `P2DENS::catalogName()` instead of hard-coded strings here? Same for viscosity below. I know you've addressed this in a comment, but I feel like this would bring some kind of closure to this huge refactoring effort. And it should be simple enough to add.

Reply:
Yes, you are right it very easy in fact. Now it looks like: ` if( strs[1] == P1DENS::catalogName() )`